OSDOCS-10877: Virtualization in Network Observability (#83105)
openshift/openshift-docs: modules/network-observability-virtualization-configuration.adoc

// Module included in the following assemblies:
//
// * network_observability/configuring-operator.adoc

:_mod-docs-content-type: PROCEDURE
[id="network-observability-virtualization-config_{context}"]
= Configuring virtual machine (VM) secondary network interfaces for Network Observability

> You can observe networking patterns on an OpenShift Virtualization setup by identifying eBPF-enriched network flows coming from VMs that are connected to additional network interfaces.

Reviewer: Should we add a mention that [default pod network](https://docs.openshift.com/container-platform/4.16/virt/vm_networking/virt-connecting-vm-to-default-pod-network.html) are covered out of the box ?

Author: @jpinsonneau do you mean that if a user has this configured, the flows from this default pod network are automatically captured by Network Observability? Or that the default pod networks are automatically configured out of the box?

// NOTE to tech review: 

> // do we need to mention anything about CNI here? I noted this in our convo, but I'm not sure if its relevant in the context of this procedure. Maybe its a prerequisite?  

Reviewer: The virt docs explicitly mention [OVN-Kubernetes secondary network](https://docs.openshift.com/container-platform/4.16/virt/vm_networking/virt-connecting-vm-to-ovn-secondary-network.html) so I would suggest to only refer to that in our doc too as it's what we tested :smile_cat:

Author: Okay, so I changed the intro sentence to:
You can observe networking patterns on an OpenShift Virtualization setup by identifying eBPF-enriched network flows coming from VMs that are connected to _an Open Virtual Network (OVN)-Kubernetes secondary network._

Reviewer 2: +1

// You can identify network flows coming from virtual machines connected to additional network interfaces that are using the Network Observability CNI plugin.

 
.Prerequisites
* Access to an {product-title} cluster with an additional network interface, such as a secondary interface or an L2 network.

> * Any other prerequisites?

Reviewer: As same as [SRIOV case](https://docs.openshift.com/container-platform/4.17/observability/network_observability/configuring-operator.html#network-observability-SR-IOV-config_network_observability), we need to have `spec.agent.ebpf.privileged` field to `true`

cc @msherif1234

Author: I have this listed as a FlowCollector config: https://github.com/openshift/openshift-docs/pull/83105/files#diff-2e10b29803e092e329b07fcd472f336c98d04f3eed531135e2afb892b7b22db0R111. I could also call it out in the prerequisites though.

Reviewer 2: yes, that's right, we should probably have dedicated menu item for "secondary interfaces" and list common pre-requisites there.

Author: Ok I added this. I feel that the intro text for "secondary interfaces" with Network Observability is lacking, and I'm wondering about potentially mixing up SR-IOV VMs and then SR-IOV interfaces. Are those two separate things? I left a //comment to you guys there trying to parse my confusion 🥴  Just trying to better understand the bigger Network Observability/secondary networks picture

  name: netdevice67619-1
  namespace: openshift-sriov-network-operator
  resourceVersion: "159492"
  uid: 13131f06-e7b1-4b42-8d95-ac7c53e908cb
spec:
  config: |-
    {
        "cniVersion": "1.0.0",
        "name": "netdevice67619-1", <1>
        "type": "sriov",
        "vlan": 0,
        "vlanQoS": 0,
        "logLevel": "info",
        "ipam": {
            "type": "static",
            "addresses": [
                {
                    "address": "192.168.122.71/24" <2>
                }
            ]
        }
----
<1> Name of the virtual virtual machine connected to the additional network interface.
<2> The MAC address to identify the virtual machine.

.. For L2, look for the annotated fields:
+
[source,yaml]
----
 # ...
  {
    "name": "ovn-kubernetes",
    "interface": "eth0",
    "ips": [
      "10.129.2.39"
    ],
    "mac": "0a:58:0a:81:02:27",
    "default": true,
    "dns": {}
  },
  {
    "name": "my-vms/l2-network",   <1>
    "interface": "podc0f69e19ba2",
    "mac": "02:fb:f8:00:00:12",    <2>
    "dns": {}
  }
 # ... 
----
<1> The name of the virtual machine connected to the additional network interface.
<2> The MAC address to identify the virtual machine.

> . Configure `FlowCollector` based on the information you found from the additional network investigation. 

Reviewer: This whole section is not needed to me.

You can directly configure the secondary network discovery without checking at these.

Reviewer: ```suggestion
. Get information about the virtual machine launcher pod by running the following command. This information is used in Step 2: 
+
[source,terminal]
----
$ oc get pod virt-launcher-<vm-name>-<random-suffix> -n <namespace> -o yaml
----
+
[source,yaml]
----
$ oc get pod virt-launcher-fedora-aqua-fowl-13-zr2x9 -n my-vms -o yaml
apiVersion: v1
kind: Pod
metadata:
  annotations:
    k8s.v1.cni.cncf.io/network-status: |- 
      [{
        "name": "ovn-kubernetes",
        "interface": "eth0",
        "ips": [
          "10.129.2.39"
        ],
        "mac": "0a:58:0a:81:02:27",
        "default": true,
        "dns": {}
      },
      {
        "name": "my-vms/l2-network",   <1>
        "interface": "podc0f69e19ba2", <2>
        "mac": "02:fb:f8:00:00:12",    <3>
        "dns": {}
      }]
  name: virt-launcher-fedora-aqua-fowl-13-zr2x9
  namespace: default
spec:
  ...
status:
  ...
----
<1> The name of the secondary network.
<2> The network interface name of the secondary network..
<3> The MAC address used for secondary network.

. Configure `FlowCollector` based on the information you found from the additional network investigation. 
```

Reviewer 2: > You can directly configure the secondary network discovery without checking at these.

just configuring name of secondary network wouldn't be enough, right? I think some sort of indexing would definitely be needed for enrichment, no?

Reviewer: Yes @memodi, you need to specify the fields that define the index. However, you don't really care about the values here.

Either we let the user check the `k8s.v1.cni.cncf.io/network-status` and figure out by themself which fields are reliable or we directly tell them for each type of secondary network which fields to use as index.

+

> [source,yaml]
> ----
> apiVersion: flows.netobserv.io/v1beta2
> kind: FlowCollector
> metadata:
>   name: cluster
> spec:
> # ...
> ebpf:
>     privileged: true       <1>
>   processor:
>     advanced:
>       secondaryNetworks:
>       - index:
>         - MAC "02:fb:f8:00:00:12"  <2>
>         - IP  "192.168.122.71/24"  <3>
>         name: my-vms/l2-network    <4>
> # ...
> ----
> <1> Ensure that the ebpf agent is in `privileged` mode so that the flows are enriched according to the MAC address.
> <2> If your additional network information has a MAC address, specify add it here.
> <3> If your additional network information has an IP address, specify add it here.
> <4> Specify the name of the virtual machine. 
> 
> . Ping from one VM to another for secondary interface IP

Reviewer: ```suggestion
[source,yaml]
----
apiVersion: flows.netobserv.io/v1beta2
kind: FlowCollector
metadata:
  name: cluster
spec:
# ...
ebpf:
    privileged: true       <1>
  processor:
    advanced:
      secondaryNetworks:
      - index: <2>
        - Interface
        - MAC
        - IP
        name: my-vms/l2-network    <3>
# ...
----
<1> Ensure that the ebpf agent is in `privileged` mode so that the flows are enriched according to the MAC address.
<2> Define the fields to use for indexing the virtual machine pods. These should form a unique Pod identifier across the cluster.
<3> Specify the name of the network. Usually <namespace>/<network-attachement-definition-name>. 
```

Reviewer 2: Discussed this with @skrthomas on slack -  that we should keep general recommendation such as indexing by MAC only. For other fields we could simply add a note saying, indexing is also supported for IP and Interface fields in combination with MAC if they’re published in annotations.

// Module included in the following assemblies:
//
// * observability/network_observability/network-observability-secondary-networks.adoc

:_mod-docs-content-type: PROCEDURE
[id="network-observability-virtualization-config_{context}"]
= Configuring virtual machine (VM) secondary network interfaces for Network Observability

> You can observe network traffic on an OpenShift Virtualization setup by identifying eBPF-enriched network flows coming from VMs that are connected to secondary networks, such as through Open Virtual Network (OVN)-Kubernetes. Network flows coming from VMs that are connected to the default internal pod network are automatically captured by Network Observability.

Reviewer 3: Same comment about "Open Virtual Network (OVN)-Kubernetes"

.Procedure
. Get information about the virtual machine launcher pod by running the following command. This information is used in Step 5: 
+
[source,terminal]
----
$ oc get pod virt-launcher-<vm_name>-<suffix> -n <namespace> -o yaml
----
+
[source,yaml]
----

> $ oc get pod virt-launcher-fedora-aqua-fowl-13-zr2x9 -n my-vms -o yaml

Reviewer 3: I'm assuming this line is supposed to be deleted since you show it in the code block above

Author: yep

apiVersion: v1
kind: Pod
metadata:
  annotations:
    k8s.v1.cni.cncf.io/network-status: |- 
      [{
        "name": "ovn-kubernetes",
        "interface": "eth0",
        "ips": [
          "10.129.2.39"
        ],
        "mac": "0a:58:0a:81:02:27",
        "default": true,
        "dns": {}
      },
      {
        "name": "my-vms/l2-network",   <1>
        "interface": "podc0f69e19ba2", <2>
        "ips": [
          "10.10.10.15"

>         ], <3>

Reviewer 3: Should this #3 callout be on line 39 or 40 instead? And indented out with the others? It gets a little lost down here now

        "mac": "02:fb:f8:00:00:12",    <4>
        "dns": {}
      }]
  name: virt-launcher-fedora-aqua-fowl-13-zr2x9
  namespace: my-vms
spec:

>   ...

Reviewer 3: These truncated ellipses should have the `#` in front of them, per https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#signs-and-symbols-in-code-blocks

status:
  ...
----
<1> The name of the secondary network.
<2> The network interface name of the secondary network.
<3> The list of IPs used by the secondary network.
<4> The MAC address used for secondary network.

. In the web console, navigate to *Operators* -> *Installed Operators*.
. Under the *Provided APIs* heading for the *NetObserv Operator*, select *Flow Collector*.
. Select *cluster* and then select the *YAML* tab.

> . Configure `FlowCollector` based on the information you found from the additional network investigation. 

Reviewer 3: ```suggestion
. Configure `FlowCollector` based on the information you found from the additional network investigation: 
```

+
[source,yaml]
----
apiVersion: flows.netobserv.io/v1beta2
kind: FlowCollector
metadata:
  name: cluster
spec:
# ...
ebpf:
    privileged: true              <1>
  processor:
    advanced:
      secondaryNetworks:
      - index:                    <2>
        - MAC                     <3>
        name: my-vms/l2-network   <4> 
# ...
----

> <.> Ensure that the eBPF agent is in `privileged` mode so that flows are collected for secondary interfaces.

Reviewer 3: Is there a reason you're not just using the numbers 1-4 for these callouts like you did above?

Author: Yes, I kept getting asciidoc warnings that my numbering was off, despite the fact that I scrutinized the numbering and it was not off. I originally had different callout/numbers and then when I changed things, I got an error. So I made them <.>

<.> Define the fields to use for indexing the virtual machine launcher pods. It is recommended to use the `MAC` address as the indexing field to get network flows enrichment for secondary interfaces. If you have overlapping MAC address between pods, then additional indexing fields, such as `IP` and `Interface`, could be added to have accurate enrichment.
<.> If your additional network information has a MAC address, add `MAC` to the field list.

> <.> Specify the name of the network found in `k8s.v1.cni.cncf.io/network-status` annotation. Usually <namespace>/<network_attachement_definition_name>. 

Reviewer 3: A possible suggestion (plus a typo)

```suggestion
<.> Specify the name of the network found in `k8s.v1.cni.cncf.io/network-status` annotation. This usually follows the pattern of `<namespace>/<network_attachment_definition_name>`. 
```

. Observe VM traffic:
.. Navigate to the *Network Traffic* page.

> .. Filter by *Source* IP using your virtual machine IP found in `k8s.v1.cni.cncf.io/network-status` annotation.

Reviewer 3: ```suggestion
.. Filter by *Source* IP using your virtual machine IP found in the `k8s.v1.cni.cncf.io/network-status` annotation.
```

> .. View both *Source* and *Destination* fields should enriched identifying VM launcher Pods and VM instance as Owners

Reviewer 3: I've re-read this sentence a couple times and still can't quite figure out what it's trying to say. I think it's missing a few words. I can't even give a suggested rewrite. I think you at least need a "the" in front of the field names. Is it trying to say to check these two fields to ensure that <something with the end of the sentence>.

Also needs a period at the end, and do "Pods" and "Owners" have to be capitalized?

Author: yeah I think its missing some words. It should be 
> View both *Source* and *Destination* fields, which should be enriched, and identify the VM launcher pods and the VM instance as owners.

// Module included in the following assemblies:
//
// * observability/network_observability/network-observability-secondary-networks.adoc

:_mod-docs-content-type: PROCEDURE
[id="network-observability-virtualization-config_{context}"]
= Configuring virtual machine (VM) secondary network interfaces for Network Observability

> You can observe networking patterns on an OpenShift Virtualization setup by identifying eBPF-enriched network flows coming from VMs that are connected to secondary networks, such as through Open Virtual Network (OVN)-Kubernetes, or SR-IOV CNI plugins. Network flows coming from VMs that are connected to the default internal pod network are automatically captured by Network Observability.

Reviewer: ```suggestion
You can observe networking patterns on an OpenShift Virtualization setup by identifying eBPF-enriched network flows coming from VMs that are connected to secondary networks. Network flows coming from VMs that are connected to the default internal pod network are automatically captured by Network Observability.
```

I would remove the mention of SRIOV here as it's not the goal in VMs

> You can observe networking patterns on an OpenShift Virtualization setup by identifying eBPF-enriched network flows coming from VMs that are connected to secondary networks, such as through Open Virtual Network (OVN)-Kubernetes, or SR-IOV CNI plugins. Network flows coming from VMs that are connected to the default internal pod network are automatically captured by Network Observability.

Author: > do we want to mention here about ebpf must be in privileged mode to be inline with what you have SRIOV? I see you have already mentioned as pre-requisite.

Since its a prerequisite and also a code callout, i think mentioning it in two places is good enough.

        ],
        "mac": "0a:58:0a:81:02:27",
        "default": true,
        "dns": {}
      },
      {
        "name": "my-vms/l2-network",   <1>
        "interface": "podc0f69e19ba2", <2>
        "ips": [
          "10.10.10.15"
        ], <3>
        "mac": "02:fb:f8:00:00:12",    <4>
        "dns": {}
      }]
  name: virt-launcher-fedora-aqua-fowl-13-zr2x9
  namespace: my-vms
spec:
  ...
status:
  ...
----
<1> The name of the secondary network.
<2> The network interface name of the secondary network.
<3> The list of IPs used by the secondary network.
<4> The MAC address used for secondary network.

. In the web console, navigate to *Operators* -> *Installed Operators*.
. Under the *Provided APIs* heading for the *NetObserv Operator*, select *Flow Collector*.
. Select *cluster* and then select the *YAML* tab.
. Configure `FlowCollector` based on the information you found from the additional network investigation. 
+
[source,yaml]
----
apiVersion: flows.netobserv.io/v1beta2
kind: FlowCollector
metadata:
  name: cluster
spec:
# ...
ebpf:
    privileged: true              <1>
  processor:
    advanced:
      secondaryNetworks:
      - index:                    <2>

>         - Interface

Reviewer: +1, it will be more obvious for the user

>         - Interface

Author: Okay I added `IP` and `Interface` as part of the callout <2> text and removed them from the example yaml + corresponding callouts.

        - MAC                     <3>
        - IP                      <4>
        name: my-vms/l2-network   <5> 
# ...
----
<.> Ensure that the ebpf agent is in `privileged` mode so that the flows are enriched according to the MAC address.

> <.> Define the fields to use for indexing the virtual machine launcher pods. These should form a unique identifier across the cluster according to the fields available in `k8s.v1.cni.cncf.io/network-status` annotation. Not all secondary networks have all the index fields in this example. Ensure that only the fields that are annotated in your secondary interface are listed here and if they are not annotated, they are removed.

Author: I kept a combo of @jpinsonneau and @memodi suggestions and added this bit:
> Not all secondary networks have all the index fields in this example. Ensure that only the fields that are annotated in your secondary interface are listed here and if they are not annotated, they are removed.

> <.> Define the fields to use for indexing the virtual machine launcher pods. These should form a unique identifier across the cluster according to the fields available in `k8s.v1.cni.cncf.io/network-status` annotation. Not all secondary networks have all the index fields in this example. Ensure that only the fields that are annotated in your secondary interface are listed here and if they are not annotated, they are removed.

Author: Addressed, thanks for the suggestion.

> <.> If your additional network information has a MAC address, specify it here.
> <.> If your additional network information has an IP address, specify it here.

Reviewer: ```suggestion
<.> If your additional network information has a MAC address, add `MAC` to the field list.
<.> If your additional network information has an IP address, add `IP` to the field list.
```

The `specify it here` mentions are confusing to me. I feel it requires to fill the values found in the pod annotation instead of the field names.

Author: Makes sense. Added your suggestion

<.> Specify the name of the network found in `k8s.v1.cni.cncf.io/network-status` annotation. Usually <namespace>/<network_attachement_definition_name>. 

> 

Reviewer 2: Even though you have omitted index by Interface here, but if users see this example and use Interface as indexing field they will not get enrichment since when the flows are captured it matches interface name partially just for the hex portion of it between what's published in annotation and what ebpf sees. Therefore it's important we only mention MAC index as an example and add a note about IP and Interface.

Author: Got it and addressed as part of your other comment on this. I was initially thinking I could try to make it clear in the callouts that they would need to be careful to omit any of the index fields that weren't relevant to their secondary interface, but I see your point about the risk if they copy our sample and use it without actually reading all the callouts.

> . Ping from one VM to another for secondary interface IP.

Reviewer: +1

> . Ping from one VM to another for secondary interface IP.

Author: Got it, will remove.

.Verification
. Navigate to the *Network Traffic* page.
. Filter by *Source* IP using your virtual machine IP found in `k8s.v1.cni.cncf.io/network-status` annotation.

> . View both *Source* and *Destination* fields should enriched identifying VM launcher Pods and VM instance as Owners

Reviewer 2: Instead of "verification" could we say something like "Observe VM Traffic" and after Navigate to the *Network Traffic* page bullet:
- filter by Source/Destination Owner name == VM name to view traffic from/to those VMs

Reviewer: This could even be a new section, more generic, giving the summary of any scenario (SRIOV & VMs)

Author: Yeah I can make it a step with substeps. For the sake of time, I'll just do that instead of making a new section for both bc that would technically need to be a new module
